app/src/main/java/com/dessalines/thumbkey/ui/components/keyboard/KeyboardScreen.kt

```diff
@@ -375,168 +374,184 @@ fun KeyboardScreen(
                                         },
-                                    dragReturnEnabled = dragReturnEnabled,
-                                    circularDragEnabled = circularDragEnabled,
-                                    clockwiseDragAction = clockwiseDragAction,
-                                    counterclockwiseDragAction = counterclockwiseDragAction,
-                                )
+                                        onAutoCapitalize = { enable ->
+                                            if (mode !== KeyboardMode.NUMERIC) {
+                                                if (enable) {
+                                                    mode = KeyboardMode.SHIFTED
+                                                } else if (!capsLock) {
+                                                    mode = KeyboardMode.MAIN
+                                                }
+                                            }
+                                        },
+                                        onSwitchLanguage = onSwitchLanguage,
+                                        onChangePosition = onChangePosition,
+                                        oppositeCaseKey =
+                                            when (mode) {
+                                                KeyboardMode.MAIN -> keyboardDefinition.modes.shifted
+                                                KeyboardMode.SHIFTED -> keyboardDefinition.modes.main
+                                                else -> null
+                                            }?.arr?.getOrNull(i)?.getOrNull(j),
+                                        numericKey =
+                                            when (mode) {
+                                                KeyboardMode.MAIN, KeyboardMode.SHIFTED, KeyboardMode.CTRLED, KeyboardMode.ALTED ->
+                                                    keyboardDefinition.modes.numeric.arr
+                                                        .getOrNull(i)
+                                                        ?.getOrNull(j)
+
+                                                else -> null
+                                            },
+                                        dragReturnEnabled = dragReturnEnabled,
+                                        circularDragEnabled = circularDragEnabled,
+                                        clockwiseDragAction = clockwiseDragAction,
+                                        counterclockwiseDragAction = counterclockwiseDragAction,
+                                    )
+                                }
                             }
                         }
                     }
                 }
             }
         }
+
+        drawKeyboard(keyboardPositionToAlignment(position), backdropEnabled)
+        if (position == KeyboardPosition.Dual) {
+            drawKeyboard(keyboardPositionToAlignment(KeyboardPosition.Right), false)
+        }
```

Comment:
Perfect, this is exactly what was needed. Tested and it works great.

I'm not a *huge* fan of internal functions or composables, but externalizing it would mean like 50 vars needing to be injected to the composable. Not worth it, and messier than what's here.